pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/scheduler/NamespaceUnloadScheduler.java

```diff
@@ -0,0 +1,155 @@
+
+    private volatile CompletableFuture<Void> currentRunningFuture = null;
+
+    public NamespaceUnloadScheduler(ScheduledExecutorService loadManagerExecutor,
+                                    LoadManagerContext context,
+                                    ServiceUnitStateChannel channel) {
+        this(loadManagerExecutor, context, channel, createNamespaceUnloadStrategy(context.brokerConfiguration()));
+    }
+
+    @VisibleForTesting
+    protected NamespaceUnloadScheduler(ScheduledExecutorService loadManagerExecutor,
+                                       LoadManagerContext context,
+                                       ServiceUnitStateChannel channel,
+                                       NamespaceUnloadStrategy strategy) {
+        this.namespaceUnloadStrategy = strategy;
+        this.recentlyUnloadedBundles = new ConcurrentHashMap<>();
+        this.recentlyUnloadedBrokers = new ConcurrentHashMap<>();
+        this.loadManagerExecutor = loadManagerExecutor;
+        this.context = context;
+        this.conf = context.brokerConfiguration();
+        this.channel = channel;
+    }
+
+    @Override
+    public synchronized void execute() {
+        boolean debugMode = conf.isLoadBalancerDebugModeEnabled() || log.isDebugEnabled();
+        if (debugMode) {
+            log.info("Load balancer enabled: {}, Shedding enabled: {}.",
+                    conf.isLoadBalancerEnabled(), conf.isLoadBalancerSheddingEnabled());
+        }
+        if (currentRunningFuture != null && !currentRunningFuture.isDone()
+                || !(conf.isLoadBalancerEnabled() && conf.isLoadBalancerSheddingEnabled())) {
+            return;
+        }
+        this.currentRunningFuture = channel.isChannelOwnerAsync().thenCompose(isChannelOwner -> {
+            if (!isChannelOwner) {
+                return CompletableFuture.completedFuture(null);
+            }
+            return context.brokerRegistry().getAvailableBrokersAsync().thenCompose(availableBrokers -> {
+                if (debugMode) {
+                   log.info("Available brokers: {}", availableBrokers);
+                }
+                if (availableBrokers.size() <= 1) {
+                    log.info("Only 1 broker available: no load shedding will be performed");
+                    return CompletableFuture.completedFuture(null);
+                }
+                final UnloadDecision unloadDecision = namespaceUnloadStrategy
+                        .findBundlesForUnloading(context, recentlyUnloadedBundles, recentlyUnloadedBrokers);
+                if (debugMode) {
+                    log.info("[{}] Unload decision result: {}",
+                            namespaceUnloadStrategy.getClass().getSimpleName(), unloadDecision.toString());
+                }
+                if (unloadDecision.getUnloads().isEmpty()) {
+                    return CompletableFuture.completedFuture(null);
+                }
+                List<CompletableFuture<Void>> futures = new ArrayList<>();
```

Comment:
Add a member var `private final Map<String, SplitDecision> inFlightUnloads` to dedup the in-flight unloads. We need to remove the item when `BSC.closeServiceUnit()` is done.

I propose moving `closeServiceUnit()` from BSC to this class(let's rename this class to UnloadManager too). This means BSC constructor should take an UnloadManager instance, as now BSC depends on `UnloadManager`.

Reply:
Here, I think we need `SplitManager and UnloadManager` to 
1. dedup inflight bundles' unload/split event pub messages.
2. move unload(closeServiceUnit) and split(splitServiceUnit) funcs outside of BSC.
3. update the success counters when the unload(closeServiceUnit) and split(splitServiceUnit) is complete.

Here is my temp code where I am trying to introduce `SplitManager` for the same reason.
https://github.com/heesung-sn/pulsar/pull/30/files#r1103463024

Reply:
Or, I am fine with moving forward with the current code(without dedup) and refactoring it later when other PRs get merged.

Reply:
> We need to remove the item when BSC.closeServiceUnit() is done.

Good catch! Since only the channel owner can do the bundle unload, removing the item when `BSC.closeServiceUnit()` might not be a good idea. Maybe we should remove it when it received an `Owned` state?

Reply:
Yes, thats better. We can add this logic in handleOwned.